### DISCUSSION TITLE
api confusion

### DISCUSSION THREAD

**Discussion by zooteld** [ggml-org/llama.cpp#3942] [Q&A] [answered]
I apologize for starting this question string in the issues. now that I am in the correct place... I am attempting to connect https://github.com/zooteld/llama-cpp-python-streamlit to my llama.cpp server running locally. I can see that something is actually hitting the server via the console, but I am not recieving anything back from the server to the app. The author of the app informed me its an endpoint issue, (uses differend json structure). it is wrote to use the llama-cpp-python bindings. any pointers on how to tackle this?

**Comment by KerfuffleV2** [comment]
Whatever sends requests to the `server` example would have to use the format that example expects. I took a very quick look at the repo you link. It seems like it may be using the OpenAI-style format. If that's the case, this might help you: https://github.com/ggerganov/llama.cpp/tree/master/examples/server#api-like-oai

I've never used it personally, so I probably won't be able to help with follow up questions on its use or anything like that.

**Comment by zooteld** [comment]
ya that is not working either... I noticed that to get it to connect I am having to change the /v1/chat/"/completion to just completion to match what is in my llama.cpp/example/server/public folder

**Reply by KerfuffleV2** [reply]
You're running it like this:

1. Start the `server` example.
2. Start the `api_like_oai` script and point it at the running `server`.
3. Run your streamlit thing and point that at whatever port `api_like_oai` is listening on.

Correct? I.E. `api_like_oai` is meant to work like a proxy and translate OpenAI-type requests into something `server` can understand.

**Reply by zooteld** [reply]
well i was starting it with the server and in my start command i have --path pointing to the examples/server/public folder so its launching the index.html

**Reply by KerfuffleV2** [reply]
It's not clear what "it" here is referring to.

To make this work you should have the server running on port 8080 for example. Then you start `api_like_OAI.py` - it will connect to 8080 by default, and listen for requests on port 8081 by default.

Therefore, your streamlit thing should be connecting to port 8081 and the paths, etc should be the same as if you were querying an OpenAI endpoint.

**Comment by zooteld** [comment]
I was referring to the api_like_OAI.py as it. lol

**Reply by KerfuffleV2** [reply]
But `api_like_OAI.py` doesn't even support a `--path` option.

**Reply by zooteld** [reply]
I believe the server does tho
here is the run command I have been using:
GGML_OPENCL_PLATFORM=AMD GGML_OPENCL_DEVICE=1 sudo ./server --alias MAPES --model models/CodeLlama-2-13b-chat-hf/ggml-model-q4_0.gguf --ctx-size 8192 --n-gpu-layers 10 --tensor-split 1.6,1.6,1.6,1.6,1.6,1.6 --batch-size 14336 --mlock --host 192.168.1.117 --port 13606 --numa --path ../../examples/server/public/api_like_OAI.py

**Reply by KerfuffleV2** [reply]
Please refer to this comment: https://github.com/ggerganov/llama.cpp/discussions/3942#discussioncomment-7477749

`api_like_OAI.py` is a script. You have to run it separately, like I said in the linked comment. You shouldn't need to specify `--path` for `server`.

The idea is to have your streamlit script connect to `api_like_OAI.py`, then `api_like_OAI.py` connects to `server` and translates the request from your streamlit thing from OpenAI-style to what `server` understands.

It's a proxy.

**Reply by zooteld** [reply]
I tried starting the api_like_OAI.py after starting the server and when i send it a request I get an error in the console: Max retries exceeded with url: /completion (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object

**Reply by zooteld** [reply]
do i need to set a key with the api_like_OAI.py?

**Reply by KerfuffleV2** [reply]
> when i send it a request I get an error in the console:

Are you still passing `--path` to `server`? If so, remove that.

> do i need to set a key with the api_like_OAI.py?

Only if you want to limit access to it, but then you'll also need to authenticate when you try to connect to that.

**Reply by zooteld** [reply]
I removed the --path part of the server launch command. I was just using that to launch the api_like_OAI.py along with the server.

**Reply by KerfuffleV2** [reply]
I'm sorry, I just don't really have the energy to keep trying to pry information out of you. I've been trying to help you, but you aren't really volunteering anything and what you've said so far is pretty vague.

Debugging issues is mostly a process of elimination, but that's only possible if the person trying to help you has pretty detailed, specific information. You take take this advice or ignore it, totally up to you but my suggestion would be to try to anticipate what the person helping you would need to know to solve the problem. And then tell them, without waiting for them to ask you. More information is pretty much always better, and the more specific it is the better as well.

Maybe someone else will be able to help you. When/if that happens, you might consider sharing the exact commands you tried to run, the output they produced, etc. I hope you will be able to resolve your issues.

**Reply by zooteld** [reply]
ah no I apologize... I'm new to the concept of using the discussions. and I'm a new kinda still learning. 

to launch the server I made a run.sh with: GGML_OPENCL_PLATFORM=AMD GGML_OPENCL_DEVICE=1 sudo ./server --alias MAPES --model models/codellama13b/ggml-model-q4_0.gguf --ctx-size 8192 --n-gpu-layers 10 --tensor-split 1.6,1.6,1.6,1.6,1.6,1.6 --batch-size 14336 --mlock --host 192.168.1.117 --port 13606

I'm running that on a "refurbished" crypto mining rig with 6 amd rx580 4gb cards in it. 

the app is on an orangepi5 on the same lan.
In the app I'm calling the http://192.168.1.117:13606/compleations
on the server starting up 'python3 api_like_OAI.py' and pointing my app at the http://192.168.1.117/completionand i get a weird exception :
Ungültiger JSON-String: <!doctype html>

Ungültiger JSON-String: <html lang=en>

Ungültiger JSON-String: <title>404 Not Found</title>

Ungültiger JSON-String: <h1>Not Found</h1>

Ungültiger JSON-String: <p>The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.</p>

I understand its most likely something simple or my "prompt format" but I haven't found much on "this does this and that does this or that" with the format it needs. but I could be just not understanding.

**Reply by zooteld** [reply]
its not a problem with lama.cpp It running and works good with the provided index.html

**Reply by KerfuffleV2** [reply]
> In the app I'm calling the http://192.168.1.117:13606/compleations

Is that actually what you used?

Because it's not the same as:

> on the server starting up 'python3 api_like_OAI.py' and pointing my app at the http://192.168.1.117/completion

Also, like I mentioned before, `api_like_OAI.py` will listen on 8081 by default. So you either set it to a different port and didn't mention it now or you're not trying to connect to the correct thing.

> sudo ./server --alias MAPES --model models/codellama13b/ggml-model-q4_0.gguf --ctx-size 8192 --n-gpu-layers 10 --tensor-split 1.6,1.6,1.6,1.6,1.6,1.6 --batch-size 14336 --mlock --host 192.168.1.117 --port 13606

Which seems to be the case. You're saying `server` is listening on `13606`, so your app should not be connecting to that. It should be connecting to whatever `api_like_OAI.py` is listening on.

`api_like_OAI.py` is what should be connecting to `server`.

It should look like this:

```plaintext
Your app - the streamlit thing.
            ||
            VV
      api_like_OAI.py
            ||
            VV
          server
```

The point of `api_like_OAI.py` is to translate requests from the OpenAI format to what `server` understands.

Just bear in mind what I said initially and that is I _think_ what you need is the OpenAI-type API but I don't know for sure. I didn't realize it was going to be such a process to reach this point. So after all this, it still might not actually work.

For `api_like_OAI.py`, `--host` and `--port` is what _it_ listens on. (`127.0.0.1:8081` by default). `--llama-api` should be what `server` is listening on. So in your case, you would want to do `--lama-api http://192.168.1.117:13606`. I am pretty sure you're not supposed to specify a path or anything, just the host and port.

**Reply by zooteld** [reply]
sorry that was a typo on that address its http://192.168.1.117:13606/compleation

**Reply by zooteld** [reply]
![2023-11-04_110654](https://github.com/ggerganov/llama.cpp/assets/22121839/3ceff48a-a650-4b98-8454-f03e0403c184)
![2023-11-05_191421](https://github.com/ggerganov/llama.cpp/assets/22121839/34d8c133-7778-4e23-9bf2-b4d890ee83eb)
![2023-11-05_191531](https://github.com/ggerganov/llama.cpp/assets/22121839/bc9068b8-b9c3-42c7-84ce-4f1e778cf81e)
![2023-11-05_191608](https://github.com/ggerganov/llama.cpp/assets/22121839/9bb2d150-e319-4e87-b7ca-57d317829372)
![2023-11-05_191634](https://github.com/ggerganov/llama.cpp/assets/22121839/b7d5d1c6-1991-4500-8331-f211c1f3088c)
![2023-11-05_191708](https://github.com/ggerganov/llama.cpp/assets/22121839/9d65e3ab-f664-4f55-bef4-cb00737ad4b1)
![2023-11-05_191735](https://github.com/ggerganov/llama.cpp/assets/22121839/1ca3c7f1-d486-4039-ab95-6c4ed8eb1e36)

**Reply by zooteld** [reply]
so ya its app>api_like_OAI.py>server

**Reply by KerfuffleV2** [reply]
It's not clear if you're saying it worked or not. You only showed output from the `server` example, and that shows requests with 200 status which indicates success.

**Reply by zooteld** [reply]
ah no Im still recieving nothing back to the app. an its still not showing a task in the console when the app requests from the api
but something is hitting the server tho

**Reply by KerfuffleV2** [reply]
You can try some other OpenAI-compatible endpoint or client to make sure that the `api_like_OAI.py` and `server` side isn't where the problem exists. If that works, then you can eliminate those as potential issues.

**Reply by S-Ali-Zaidi** [reply]
I had this same issue using api_like_OAI.py with the ChatBox program on my Mac. The endpoint at port 8081 using api_like_OAI.py is able to stream text to SOME UI programs like MacGPT, but it’s unable to do so with Chatbox. 

But Chatbox is able to receive and display text that is sent using a server created by Studio LM (which is built on llama.cpp) as well as the OpenAI servers.. So not just why this is happening. I tried to Wireshark the outgoing chunks from both api_like_OAI.py and StudioLM and couldn’t find any difference that would break streaming functionality for some apps.

**Reply by KerfuffleV2** [reply]
You could possibly try #3972, it says it's fixing some `api_like_OAI.py` issues.

**Comment by zooteld** [comment, marked as answer]
for my use in the app I was beating my head against the desk trying to get working, the author of the origional version pushed an update enabling endpoint swapping. So now after launching the server and ai_like_OAI.py, the app connects to 8081 and uses /v1/completions as an  endpoint. thank you all and especially KerfuffleV2 for putting up with me.

**Reply by KerfuffleV2** [reply]
No problem. Glad you managed to solve your problem!
